### PR TITLE
Redefine IS_LAYER_ON/OFF() as aliases for existing layer functions

### PR DESCRIPTION
**Pull request by vomindoraan**
## Description

Redefines `IS_LAYER_ON()` and `IS_LAYER_OFF()` macros as aliases for the core function `layer_state_is()`. This function covers an edge case that the existing macros do not (see [here](https://github.com/qmk/qmk_firmware/pull/6352#issuecomment-605524047)).

Also adds `IS_LAYER_ON_STATE()` and `IS_LAYER_OFF_STATE()` as aliases for `layer_state_cmp()`. These macros work like `IS_LAYER_ON()` and `IS_LAYER_OFF()`, respectively, except they allow the layer state to be parametrized instead of always using the global `layer_state`.

Together these four macros are analogous to the `IS_HOST_LED_ON()` and `IS_LED_ON()` macros (added in #4853), except that they check layer state instead of LED state.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
We actually already have a function that does this: `layer_state_cmp`. 

https://github.com/qmk/qmk_firmware/blob/ba26736d7e0679e87e34add2445cdae541e2a0b8/tmk_core/common/action_layer.c#L154-L157

File: quantum/quantum.h [hunk 1]
```diff
@@ -219,8 +219,11 @@ void set_single_persistent_default_layer(uint8_t default_layer);
 
 void tap_random_base64(void);
 
-#define IS_LAYER_ON(layer)  (layer_state & (1UL << (layer)))
-#define IS_LAYER_OFF(layer) (~layer_state & (1UL << (layer)))
+#define IS_LAYER_ON_STATE(state, layer)  ( (state) & (1UL << (layer)))
```
Comment (drashna) [inline quantum/quantum.h]:
```suggestion
#define IS_LAYER_ON_STATE(state, layer)  layer_state_cmp(state, layer)
```

File: quantum/quantum.h [hunk 2]
```diff
@@ -219,8 +219,11 @@ void set_single_persistent_default_layer(uint8_t default_layer);
...
+#define IS_LAYER_OFF_STATE(state, layer) (~(state) & (1UL << (layer)))
```
Comment (drashna) [inline quantum/quantum.h]:
```suggestion
#define IS_LAYER_OFF_STATE(state, layer) !layer_state_cmp(state, layer)
```

File: quantum/quantum.h [hunk 3]
```diff
@@ -219,8 +219,11 @@ void set_single_persistent_default_layer(uint8_t default_layer);
...
+
+#define IS_LAYER_ON(layer)  IS_LAYER_ON_STATE(layer_state, layer)
```
Comment (drashna) [inline quantum/quantum.h]:
```suggestion
#define IS_LAYER_ON(layer)  layer_state_cmp(layer_state, layer)
```

File: quantum/quantum.h [hunk 4]
```diff
@@ -219,8 +219,11 @@ void set_single_persistent_default_layer(uint8_t default_layer);
...
+#define IS_LAYER_OFF(layer) IS_LAYER_OFF_STATE(layer_state, layer)
```
Comment (drashna) [inline quantum/quantum.h]:
```suggestion
#define IS_LAYER_OFF(layer) !layer_state_cmp(layer_state, layer)
```

**Comment by vomindoraan** [comment]
In that case, it's better to keep the macros as they currently are, because they're functionally identical to `layer_state_cmp` (the extra `if` check in the function doesn't do anything), and it's good to avoid unnecessary function calls.

**Comment by vomindoraan** [comment]
I'd even argue that `layer_state_cmp()` doesn't work correctly — if `layer_state` is somehow 0 (all layers are off), I'd expect `IS_LAYER_ON(0)` to return false, not true. But it will return true if we implement the macros using `layer_state_cmp()`.

**Comment by drashna** [comment]
I would say that it would be better to fix the `layer_state_cmp` option, in this case, actually.

**Comment by vomindoraan** [comment]
This seems like a function that was written for a fairly specific use case (the doc comment mentions unit tests). Perhaps the `if` behavior was intended, so we don't know if and what would break if we changed it. I maintain that we should keep separate quantum macros for checking layer state (as we've been doing with `IS_LAYER_ON()` thus far).

**Comment by drashna** [comment]
See, I disagree, with that.  Part of the reason that it isn't much used is that it's not documented.  

And I don't like that this adds multiple ways to do the same thing.  And not consistently.  I'd rather that there be just a single function for this.  But that's my opinion.

**Comment by vomindoraan** [comment]
Could anyone else (@noroadsleft @fauxpark @zvecr) chime in on the “keep macros separate” vs “change how the function works” issue that Drashna and I discussed above?

**Comment by noroadsleft** [comment]
I don't like the idea of having two different ways to do the same thing unless it's deemed absolutely necessary, because it's less maintainable.

If the `layer_state_cmp()` function is bugged it should be fixed IMO. What @drashna said about this functionality not being documented is actually the exact reason I don't use any layer checking functionality. I have need of it, but I'm not a C programmer so I have no idea what's going on.

I also feel like I lack context for this. I don't understand the relevance of needing or not needing a parameterized `state`.

**Comment by vomindoraan** [comment]
Thanks for your input. I'd argue that maintainability isn't an issue here as both the functions and macros are essentially just a single expression that is unlikely to ever change. There is only one straightforward way to check if a certain bit is set, and that is `a & (1 << b)`.

On the topic of the function being bugged: it may be, but the current behavior may be intended, and changing it, or the `layer_state_is()` function that corresponds to `IS_LAYER_ON()`, could break tests or someone's keymap. I think the `IS_LAYER_ON()`/`IS_LAYER_ON_STATE()` macros and `layer_state_is()`/`layer_state_cmp()` functions may actually serve slightly different purposes. That's why I'm suggesting that we add `IS_LAYER_ON_STATE()` and document it along with `IS_LAYER_ON()`, so they can become _the_ recommended way to do layer checks, without relying on the `tmk_core` stuff.

**Comment by vomindoraan** [comment]
Resolved conflict, updated to latest master. Also bump.

**Comment by drashna** [comment]
> It's mainly for the `layer_state_set_*()` functions, where `state` is passed as a parameter, and so the current global `layer_state` shouldn't be implicitly used.

And it is for this exact reason that I would prefer `layer_state` named functions.  Also, I'm not exactly a fan of the upper cased names here, either.

**Comment by vomindoraan** [comment]
> And it is for this exact reason that I would prefer `layer_state` named functions.

I'm not sure I understand what you mean.

> Also, I'm not exactly a fan of the upper cased names here, either.

The uppercase names are for consistency with all the existing macros: `IS_LAYER_ON()`, `IS_LAYER_OFF()`, `IS_LED_ON()`, `IS_LED_OFF()`, `IS_HOST_LED_ON()`, `IS_HOST_LED_OFF()`. Also, upper/constant-case names for macros are an established best practice.

**Comment by drashna** [comment]
I guess, mostly, it's preference/OCD for me.

**Comment by vomindoraan** [comment]
Rebased onto latest master and formatted the changes with clang-format.

**Comment by vomindoraan** [comment]
Looks like this was approved but was never merge. I still find that there's a need for this change. Could this PR be reopened?

**Comment by drashna** [comment]
Actually digging into this more, .... I have to double down on the "we should use `layer_state_cmp(state, layer)` and `layer_state_is(state)` usage of this.   In fact, I'd rather deprecate the `IS_LAYER_ON/OFF` macros.

The `layer_state_cmp` has some additional handling.  Mostly, that is so that it returns true of all layers are off, including layer 0.  And considering that layer 0 is "always active", from a functionality standpoint, using `IS_LAYER_*` could lead to situations where you get a false back, when you **_should_** get a true.

**Comment by vomindoraan** [comment]
I see what you mean. Is there a valid scenario where layer 0 could get turned off?

File: quantum/quantum.h [hunk 5]
```diff
@@ -245,8 +245,11 @@ void set_single_persistent_default_layer(uint8_t default_layer);
 
 void tap_random_base64(void);
 
-#define IS_LAYER_ON(layer) (layer_state & (1UL << (layer)))
-#define IS_LAYER_OFF(layer) (~layer_state & (1UL << (layer)))
+#define IS_LAYER_ON_STATE(state, layer) ((state) & (1UL << (layer)))
+#define IS_LAYER_OFF_STATE(state, layer) (~(state) & (1UL << (layer)))
+
+#define IS_LAYER_ON(layer) IS_LAYER_ON_STATE(layer_state, layer)
+#define IS_LAYER_OFF(layer) IS_LAYER_OFF_STATE(layer_state, layer)
```
Comment (vomindoraan) [inline quantum/quantum.h]:
```suggestion
#define IS_LAYER_ON(layer) layer_state_is(layer)
#define IS_LAYER_OFF(layer) !layer_state_is(layer)

#define IS_LAYER_ON_STATE(state, layer) layer_state_cmp(state, layer)
#define IS_LAYER_OFF_STATE(state, layer) !layer_state_cmp(state, layer)
```

**Comment by vomindoraan** [comment]
I've changed the macros into aliases for `layer_state_cmp()` and `layer_state_is()`. This now needs to be tested to check if it breaks existing `IS_LAYER_ON/OFF()` usage in certain edge cases (namely calling the macros with 0).

**Comment by drashna** [comment]
Deprecating them would be a breaking change, as a bunch of keymaps use them.  But aliasing them would be a good idea. 

As for the layer state, itself. Yes, it starts off as 0 for the bitmask, meaning that layer 0 is off by default. But the default layer is always correctly set, on startup, and the key<->layer check `OR`s the default and regular layer state.  

Though, for the documentation, it may be simpler to add a column in the table for aliases, rather than adding additional rows

**Comment by vomindoraan** [comment]
You're right, I'll update the documentation.

Thoughts on `IS_LAYER_IS_ON_STATE(state, layer)` vs `IS_STATE_LAYER_ON(state, layer)`?

**Comment by drashna** [comment]
I would personal prefer `IS_LAYER_STATE_*`, but ... I may not be the best person to ask about naming stuff.

**Comment by noroadsleft** [comment]
I agree with drashna; I think `IS_LAYER_STATE_(ON|OFF)` is more clear.

**Comment by vomindoraan** [comment]
I think `IS_LAYER_STATE_ON/OFF` might not be clear enough: to me this sounds like checking if the _layer's state_ is on/off, which is the same thing as checking if the _layer_ itself is on/off (`IS_LAYER_ON/OFF`).

I'd like there to be a distinction between what's used to check the global layer state, and what's used to check a layer in another, distinct state that's passed in as an argument. That's why the `_STATE` suffix makes sense (indicating that this version of the macro is different from the regular one with respect to state), and why `STATE_LAYER` would make sense too (indicating that the layer belongs to a particular state, not the global one).

I think I would keep it as is, with the `_STATE` suffix, but if you'd still like me to change it to `IS_LAYER_STATE_ON/OFF`, I can do that.

**Comment by vomindoraan** [comment]
Requesting re-review since the goal of the PR has changed since it was re-opened.

**Comment by drashna** [comment]
The naming is tricky here, given existing naming convention for layer stuff, and what this is doing.  That's why I made the comment earlier about naming stuff.

**Comment by vomindoraan** [comment]
I've checked all 113 user/keymap files in the repo that use `IS_LAYER_ON/OFF()`, and none of them call it with 0 (this is to be expected). I think it's safe to say that this change isn't going to break anything.

**Comment by zvecr** [comment]
> Update on testing: I checked all 113 user/keymap files in the repo that use IS_LAYER_ON/OFF(), and none of them call it with values that are 0 (this was to be expected). I think it's safe to conclude that this change isn't going to break anything

In some cases, we also potentially care about out of repo breakages. Due to that, this one would be difficult to quantify as zero risk.

**Comment by vomindoraan** [comment]
Indeed, it would be hard to say that for any change that touches something in core in this way, as technically anything could exist outside of the repo. But it is unlikely, given these numbers.

**Comment by vomindoraan** [comment]
I don't think this is awaiting any changes, just needs to be reviewed.

If it is indeed deemed a breaking change, I don't mind changing the PR to target `future`, but I don't think it's necessary for such a small change.
